### PR TITLE
Fix malloc/free mismatch identified under valgrind+sysalloc

### PR DESCRIPTION
**Pull request by akshatsikarwar**

### CODE REVIEW

File: bb/sbuf2.c [hunk 1]
```diff
@@ -716,35 +719,50 @@ void SBUF2_FUNC(sbuf2setflags)(SBUF2 *sb, int flags)
...
-        sb->fd = fd;
-        sb->flags = flags;
+#if SBUF2_SERVER
+    comdb2ma alloc = comdb2ma_create(0, 0, "sbuf2", 0);
+    if (alloc == NULL) {
+        goto error;
+    }
+    /* get malloc to work in server-mode */
+    SBUF2 dummy = {.allocator = alloc};
+    sb = &dummy;
```
Comment (riverszhang89) [inline bb/sbuf2.c]: What is this `dummy` for? If this is to make it easy to destroy the allocator upon error, we would still lose the reference to the allocator if the following malloc failed.

Comment (akshatsikarwar) [inline bb/sbuf2.c]: In server mode malloc -> comdb2_malloc(sb->allocator,...)
We create allocator first, then allocate sb from it. This sets up pointer to make above call work.

Comment (akshatsikarwar) [inline bb/sbuf2.c]: Reference is not lost - its in `alloc`

File: bb/sbuf2.c [hunk 2]
```diff
@@ -120,9 +120,12 @@ int SBUF2_FUNC(sbuf2free)(SBUF2 *sb)
         sb->dbgout = NULL;
     }
 #if SBUF2_SERVER
-    comdb2ma_destroy(sb->allocator);
+    comdb2ma alloc = sb->allocator;
 #endif
     free(sb);
+#if SBUF2_SERVER
+    comdb2ma_destroy(alloc);
```
Comment (akshatsikarwar) [inline bb/sbuf2.c]: Because `sb` is allocated from mspace.

File: bb/sbuf2.c (hunk 1 above)
Comment (akshatsikarwar) [inline bb/sbuf2.c]: Too ugly?

Comment (riverszhang89) [inline bb/sbuf2.c]: Got it!! Thanks

Comment (akshatsikarwar) [inline bb/sbuf2.c]: Didn't want to have any raw comdb2_malloc/comdb2_free calls..

**Comment by akshatsikarwar** [comment]
From logs:
`!socksql: setup failed`

From setup log, not sure what didn't work. In anycase, definitely unrelated to this commit.

File: bb/sbuf2.c (hunk 1 above)
Comment (riverszhang89) [inline bb/sbuf2.c]: No. It's quite clever.

**Comment by riverszhang89** [comment]
Awesome - Good to merge.

**Pull request merged**